neuralmonkey/encoders/cnn_encoder.py

```diff
@@ -99,11 +105,14 @@ def plain_convolution(
             prev_mask: tf.Tensor,
             specification: ConvSpec,
             layer_num: int) -> Tuple[tf.Tensor, tf.Tensor, int]:
-        if len(specification) != 5:
+        try:
+            check_argument_types()
+        except TypeError:
             raise ValueError((
```

Comment:
nerajzuješ valueerror, ale typeerror

Reply:
To schválně. Podle mě, když se spleteš v dlouhym seznamu seznamu konfiguračních tuplů (podívej se třeba do `str.ini`, experimentech používám asi 20 vrstev), že někde zapomeneš jedno číslo, tak jsi spíš udělal `ValueError` než `TypeError`.

Reply:
No ne, tohle je určitě type error protože nesedí typ.

Navíc teda by se mělo psát:
```
except TypeError as err:
  raise TypeError("message") from err
```

Reply:
No tady to je typ, na tomhle místě v kódu, ale když to píšeš v tom konfiguráku, tak ti rozhodně nepřipadá jako že ses splet v typu, ale že jsi tam napsal dlouhej string napsal nějak blbě. `TypeError` je potom matoucí, protože přece vidíš, že jsi napsal list tuplů.

Reply:
Proto je tam ta message, těžko budeš debugovat podle typu vyhozený výjimky..  Tady je to typ, a proto by měl být vyhozen `TypeError`

Reply:
Jinak ještě, když napíšu jiný písmenko do toho tuplu, třeba "R" jako Residual, ale chci "C" jako convolution a napíšu tam k tomu správně i ty argumenty, tak jsem se taky dopustil `TypeErroru`?

Navíc, když jsi zvyklej na obyčejnej Python a nepoužíváš mypy, tak by tě ani ve snu nenapadlo, že tuple (int, int) je jinej typ, než (str, str) a to víc je TypeError matoucí.

Reply:
Já nevim.

Blbý písmenko je určitě `ValueError`, ale záleží, co předpokládáš a co kontroluješ. Považuješ-li písmenko za věc, která je vždycky dobře, tak je to `TypeError`. Pokud se rozhoduješ podle toho, kolik prvků ten tupl má mít a používáš typeguard, tak je to `ValueError`

Neřikám, že to neni divný.

Když přijde někdo mypem nepolíbenej, tak se zblázní dřív, než to dočte sem.

Pořád trvám na tom, že jako uživatel když debuguju, tak si první přečtu message k tý výjimce a pokud je srozumitelná, tak je mi to, jestli je to `TypeError` nebo `ValueError` celkem jedno.

Reply:
Ať už tam bude co chce, rozhodně chci, aby tam bylo `from err`. Ten typeguard ti totiž přesně řekne, co je s tim tuplem špatně.